horovod/ray/runner.py

```diff
@@ -0,0 +1,449 @@
+        # CUDA will be able to detect adjacent devices and use IPC
+        # allowing for better performance.
+        for worker in self.workers:
+            worker.update_env_vars.remote({"CUDA_VISIBLE_DEVICES": all_ids})
+
+        ray.get([
+            worker.start_executable.remote(executable_cls, executable_args,
+                                           executable_kwargs)
+            for worker in self.workers
+        ])
+        return node_id
+
+    def get_workers(self) -> List:
+        return self.workers
+
+
+class Coordinator:
+    """Responsible for instantiating the Rendezvous server.
+
+    Args:
+        settings: Horovod Settings object."""
+    rendezvous = None
+    global_rendezv_port = None
+    nics = None
+    hostnames = None
+
+    def __init__(
+            self,
+            settings,
+    ):
+        self.settings = settings
+        self.hostnames_by_rank = defaultdict(list)
+
+    @property
+    def world_size(self) -> int:
+        return sum(len(ranks) for ranks in self.hostnames_by_rank.values())
+
+    @property
+    def hoststring(self) -> str:
+        return ",".join([
+            f"{host}:{len(ranks)}"
+            for host, ranks in self.hostnames_by_rank.items()
+        ])
+
+    def register(self, hostname: str, world_rank: int):
+        self.hostnames_by_rank[hostname].append(world_rank)
+
+    def finalize_registration(self) -> dict:
+        """Return a dictionary for all ranks."""
+        rank_to_info = {}
+        for node_world_rank, (hostname, ranks) in enumerate(
+                self.hostnames_by_rank.items()):
+            for local_rank, world_rank in enumerate(ranks):
+                rank_to_info[world_rank] = dict(
+                    NODE_WORLD_RANK=node_world_rank,
+                    NODE_WORLD_SIZE=len(self.hostnames_by_rank),
+                    LOCAL_RANK=local_rank,
+                    LOCAL_SIZE=len(ranks))
+        return rank_to_info
+
+    def establish_rendezvous(self) -> Dict[str, str]:
+        """Creates the rendezvous server and identifies the nics to be used.
+
+        Returns:
+            Environment variables for each worker.
+        """
+
+        # start global rendezvous server and get port that it is listening on
+        self.rendezvous = RendezvousServer(self.settings.verbose)
+
+        # allocate processes into slots
+        # hosts = parse_hosts(hosts_string="10.11.11.11:4,10.11.11.12:4")
+        parsed_hosts = hosts.parse_hosts(hosts_string=self.hoststring)
+        host_alloc_plan = hosts.get_host_assignments(parsed_hosts,
+                                                     self.world_size)
+
+        # start global rendezvous server and get port that it is listening on
+        self.global_rendezv_port = self.rendezvous.start()
+        self.rendezvous.init(host_alloc_plan)
+        # remote_host_names = network.filter_local_addresses()
+        self.nics = driver_service.get_common_interfaces(
```

Comment:
Will this `driver_service` requires ssh acces to each ray nodes without prompt?

Reply:
You can pass in an identity file now. Otherwise you need passwordless ssh access.

Reply:
Thank you @richardliaw for your timely responses.  

However, is this "get_common_interfaces" strictly necessary? Also asking @tgaddair .  And is it possible to set the "HOROVOD_GLOO_IFACE" variable separately on each node using the nic associated the IP returned by `ray.service.get_ip_address`?

We have tried this at https://github.com/intel-analytics/analytics-zoo to avoid the ssh requirements and it works on our environment, but I am not sure this approach is strictly correct.

Reply:
@yangw1234 can you make a new issue for this? It is hard to track comments on a closed PR! Thanks!

Reply:
Sure. Added a new issue. #2271